### PR TITLE
Validation files

### PR DESCRIPTION
**Pull request by yael1994**

### CODE REVIEW

**Comment by yael1994** [comment]
https://trello.com/c/4Q4wsRjb/14-input-files-validation

**Review by shacharmo — CHANGES_REQUESTED** [review]
Do we need samplename2biologicalcondition.txt?

Code is not connected to the pipeline

File: tools/validation_files.py [hunk 1]
```diff
@@ -1,46 +1,50 @@
...
+    print(f'barcode2samples_valid: {barcode2samples_valid} ')
+    if samples2bc_path and barcode2samples_path:
+        if samples2bc_valid and barcode2samples_valid:
+            return is_same_samples(samples2bc_path, barcode2samples_path)
+        else:
+             False
     else:
-        return False
\ No newline at end of file
+        return samples2bc_valid or barcode2samples_valid
```
Comment (shacharmo) [inline tools/validation_files.py]: shouldn't be `and` instead of `or`?

File: tools/validation_files.py [hunk 2]
```diff
@@ -1,46 +1,50 @@
...
+                find_tab = False
+                count_tab = 0
             elif num<48 or (num>57 and num<65) or (num>90 and num<95) or (num>95 and num<97) or (num>122):
                 #48-57 number, 65-90 Big letter, 97-122 small letter, 95 is _
                 return False
-            b=f.read(1)
+            b = f.read(1)
     return True        
 
 def is_validation_files(samples2bc_path, barcode2samples_path):
```
Comment (shacharmo) [inline tools/validation_files.py]: shouldn't the name be `is_input_files_valid` or something similar?
`is_validaiton_files` should like the function return whether we should validate the files and not if the files are valid

File: IgOmeProfiling_pipeline.py [hunk 3]
```diff
@@ -18,6 +19,12 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
                  allowed_gap_frequency, concurrent_cutoffs, meme_split_size, number_of_random_pssms,
                  rank_method, tfidf_method, tfidf_factor, shuffles,
                  run_summary_path, error_path, queue, verbose, argv):
+    
+    # check the validation of files barcode2samplename_path and samplename2biologicalcondition_path
+    check_files = is_validation_files(samples2bc_path=samplename2biologicalcondition_path,barcode2samples_path=barcode2samplename_path)
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: `check_files` is not a  good name, the function returns if the files are valid and not if we should check the files.
Also, we should differentiate between the different errors and log them correctly.
We don't want the user to guess which file is invalid and why

File: IgOmeProfiling_pipeline.py [hunk 4]
```diff
@@ -42,8 +49,9 @@ def run_pipeline(fastq_path, barcode2samplename_path, samplename2biologicalcondi
 
         module_parameters = [fastq_path, first_phase_output_path, first_phase_logs_path,
                              barcode2samplename_path, left_construct, right_construct,
-                             max_mismatches_allowed, min_sequencing_quality, first_phase_done_path,
-                             '--gz' if gz else '', f'--error_path {error_path}', '-v' if verbose else '']
+                             max_mismatches_allowed, min_sequencing_quality, first_phase_done_path, 
+                             '--check_files' if check_files else '', '--gz' if gz else '',
```
Comment (shacharmo) [inline IgOmeProfiling_pipeline.py]: If we already know the files are valid why are we asking other parts of the pipeline to recheck it? The name is confusing.

Each module wrapper should have a flag whether to validate files.
The default can be true and if validated the files in main pipeline then value should be false

File: tools/validation_files.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    samples_from_sample2bc = sorted(load_table_to_dict(samples2bc_path,'Samples {} belongs to more than one bc!!').keys()) if file_extension=='.txt' else []
+    #else: #json file
+        #TODO call to function that return dict from json file
+    samples_from_barcode2sample = sorted(load_table_to_dict(barcode2samples_path,'Barcode {} belongs to more than one sample!!').values())
+    return samples_from_sample2bc==samples_from_barcode2sample
+
+def is_valid_structure_file(file_path):
+    # verify the structure of the file -  should look like: word\tword\n
+    # word can conation: number, big or little letter and _
+    with open(file_path, "rb") as f:
```
Comment (shacharmo) [inline tools/validation_files.py]: All this code can be reduced the few line using regular expression

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: tools/validation_files.py (hunk 1 above)
Comment (yael1994) [reply tools/validation_files.py]: no, it is a case when there is only one file as input so it should be at least one True

Comment (shacharmo) [reply tools/validation_files.py:46]: But if we have both file and one if invalid the function will return that all is valid...

File: auxiliaries/pipeline_auxiliaries.py [hunk 6]
```diff
@@ -279,16 +280,26 @@ def fetch_cmd(script_name, parameters, verbose, error_path, done_path=None):
...
+                    assert False, error_msg.replace('{}', key)  # TODO: write to a global error log file
+                table[key] = value
+    else:
+        f = open(table_path, 'r')
+        dict_table = json.load(f)
+        for key in dict_table:
+            for val in dict_table[key]:
+                if val in table:
+                    assert False, error_msg.replace('{}', val)  # TODO: write to a global error log file
+                table[val] = key
```
Comment (shacharmo) [inline auxiliaries/pipeline_auxiliaries.py]: The validation here is not clear or correct.
We want to make sure the schema/data structure as what we expect.
The JSON files are more complicated that the simple text

File: tools/validation_files.py [hunk 7]
```diff
@@ -1,50 +1,46 @@
...
-                count_tab += 1
-                find_tab = True
-            elif num==10 and find_tab and count_tab==1: #10 is new line
-                find_tab = False
-                count_tab = 0
-            elif num<48 or (num>57 and num<65) or (num>90 and num<95) or (num>95 and num<97) or (num>122):
-                #48-57 number, 65-90 Big letter, 97-122 small letter, 95 is _
-                return False
-            b = f.read(1)
+    regex = '\w*\t\w*\n'
```
Comment (shacharmo) [inline tools/validation_files.py]: This assume files are always text but they can be JSON

File: tools/validation_files.py (hunk 1 above)
Comment (shacharmo) [reply tools/validation_files.py]: Code should be clearer/simpler to follow.
Prefer early return, i.e., in `if samples2bc_path:` you check about `barcode2samples_path` value and if empty return and vice-versa.

**Review by shacharmo — CHANGES_REQUESTED** [review]
See also the unresolved conversion regarding

File: tools/validation_files.py (hunk 1 above)
Comment (yael1994) [reply tools/validation_files.py]: but most of the time the input for the function is only one file and not two of them so if there is no file I can't do return.

File: auxiliaries/pipeline_auxiliaries.py [hunk 8]
```diff
@@ -32,6 +32,18 @@
                          "TGG": "W",
                          "TAT": "Y"}
 
+
+sample2bc_schema = {
+    "type": "object",
+    "propertyNames": {
+        "pattern": "^[A-Za-z0-9_]*$"
```
Comment (shacharmo) [inline auxiliaries/pipeline_auxiliaries.py]: `*` => `+`
or `{min_length,}`

We want at least one or few chars

File: auxiliaries/pipeline_auxiliaries.py [hunk 9]
```diff
@@ -32,6 +32,18 @@
...
+    },
+    "patternProperties": {
+        "^[A-Za-z0-9_]*$" : { "type": "array", "items": {"type": "string", "pattern": "^[A-Za-z0-9_]*$"} }
```
Comment (shacharmo) [inline auxiliaries/pipeline_auxiliaries.py]: Same comment about `*`

File: tools/validation_files.py [hunk 10]
```diff
@@ -1,46 +1,53 @@
...
+import jsonschema
+
 
 def is_same_samples(samples2bc_path, barcode2samples_path):
     # verify that file sample2biologicalcondition and file barcode2sample have the same samples.
     samples_from_sample2bc = sorted(load_table_to_dict(samples2bc_path,'Samples {} belongs to more than one bc!!').keys())
     samples_from_barcode2sample = sorted(load_table_to_dict(barcode2samples_path,'Barcode {} belongs to more than one sample!!').values())
-    return samples_from_sample2bc == samples_from_barcode2sample
+    if not samples_from_sample2bc == samples_from_barcode2sample:
+        print(f'The files: {samples2bc_path} and {barcode2samples_path} not contain the same samples!!')
```
Comment (shacharmo) [inline tools/validation_files.py]: Prefer logger.
This is applicable to all prints

File: tools/validation_files.py [hunk 11]
```diff
@@ -1,46 +1,53 @@
...
 
-    if samples2bc_path and barcode2samples_path:
-        if samples2bc_valid and barcode2samples_valid:
-            same_samples = is_same_samples(samples2bc_path, barcode2samples_path)
-            if not same_samples:
-                print(f'The files: {samples2bc_path} and {barcode2samples_path} not contain the same samples!!')
-                return False
-        else:
-            False
+def is_input_files_valid(samplename2biologicalcondition_path, barcode2samplename_path):
```
Comment (shacharmo) [inline tools/validation_files.py]: Please refactor the method it is not clear enough and prone to error.
For example file can invalid but you still test if same samples.
Hard to follow what happens if only file is supplied.
Data is loaded multiple times (validation and same sample).
Return condition doesn't follow best practice where validation functions return a single True value at the end and False value in conditions.

The following is the psuedo code for the functions:
* Set validation flags to **True**
* If bar2sam path is defined
  * Try to load data, if fails set flag to false, log and skip next line
  * Validate data, if invalid set flag to false and log
* If sam2bc path is defined, do the same as above (update the other flag)
* If any flag is false, return False (no reason to check containment)
* if both path exists and not is_same_sample: return False
* return True

File: tools/validation_files.py [hunk 12]
```diff
@@ -1,46 +1,53 @@
...
+    if barcode2samplename_path:
+        barcode2samples_valid = is_valid_txt_file(barcode2samplename_path) 
+    
+    # If there are two files - test that the two files have the same samples.
+    if (samplename2biologicalcondition_path and barcode2samplename_path):
+        return is_same_samples(samplename2biologicalcondition_path, barcode2samplename_path) if (samples2bc_valid and barcode2samples_valid) else False
     else:
-        return samples2bc_valid or barcode2samples_valid
\ No newline at end of file
+        return (samples2bc_valid or barcode2samples_valid)
```
Comment (shacharmo) [inline tools/validation_files.py]: Always add blank line at end a file (best practice)

**Review by shacharmo — CHANGES_REQUESTED** [review]

File: tools/validation_files.py [hunk 13]
```diff
@@ -1,53 +1,72 @@
 import sys
-from auxiliaries.pipeline_auxiliaries import load_table_to_dict, sample2bc_schema
+from auxiliaries.pipeline_auxiliaries import load_table_to_dict
 import os
 import re
 import json
 import jsonschema
+import datetime
+
+def is_valid_Json_file(json_path, schema, logger):
```
Comment (shacharmo) [inline tools/validation_files.py]: Function is not called anywhere

File: tools/validation_files.py [hunk 14]
```diff
@@ -1,53 +1,72 @@
...
+def is_valid_data(path_file ,dict_data, logger):
     # verify the structure of the file -  should look like: word\tword\n
-    regex = '\w*\t\w*\n'
-    for data in  open(file_path, "rb"):
-        line = data.decode("utf-8")
-        m = re.match(regex, line)
-        if m == None:
-            print(f'The structure of file name {file_path} is not valid')
+    regex = "^[A-Za-z0-9_]+$"
+    for key, value in  dict_data.items():
```
Comment (shacharmo) [inline tools/validation_files.py]: You assume JSON has the required structure but it is not validated anywhere

**Review by shacharmo — CHANGES_REQUESTED** [review]
Fix conflicts and comments below

File: tools/validation_files.py (hunk 13 above)
Comment (yael1994) [reply tools/validation_files.py:9]: I gonna use it to check the validation of the new JSON files - like the JSON for phase 1. 
That is why I put it in the share file, I thought it belongs there.
